### ISSUE TITLE
Ability to download all library and asset files for all available versions of the game

### ISSUE THREAD

**Issue by TheMCNerd2017** [PolyMC/PolyMC#403]
### Role

Wanting to have a complete set of resource files for the game after seeing Meri's YouTube video.

### Suggestion

Add the ability to download library and asset files for all available versions of the game.

### Benefit

So that players can play any version of Minecraft in the future without having to rely on Microsoft.

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

I recently saw Meri's YouTube video urging Java Edition players to download all Minecraft Java Edition resource files if they can do a little bit of scripting, and thought that having the ability to do this in PolyMC would be incredibly useful, especially since you would be able to play any version of Minecraft that was available at the time of downloading the resource files without having to rely on Microsoft.

**Comment by txtsd** [comment]
That's going to amount to a LOT of space on your system. I'm not sure we'll implement something like this.

Link to video?

**Comment by TheMCNerd2017** [comment]
> That's going to amount to a LOT of space on your system. I'm not sure we'll implement something like this.
> 
> Link to video?

Sure I'll link you to the video: https://www.youtube.com/watch?v=ymwvTTmPBI8

A part 2 to that video is also going to be uploaded in the future, but I'm not sure when.

As for it being a lot of space, apparently all of the libraries and assets would be around 23GB according to Meri's video (a screenshot is shown in the video). On my system I have lots of storage available, to the point where 23GB is not much at all. I do understand that it's not the case for everyone else though.

**Comment by DioEgizio** [comment]
> So that players can play any version of Minecraft in the future without having to rely on Microsoft.

wouldn't that be piracy?

**Comment by Scrumplex** [comment]
> > So that players can play any version of Minecraft in the future without having to rely on Microsoft.
> 
> wouldn't that be piracy?

Nope. You need a valid account to be able to download library files and assets.

**Comment by TheMCNerd2017** [comment]
> > So that players can play any version of Minecraft in the future without having to rely on Microsoft.
> 
> wouldn't that be piracy?

By relying off of Microsoft I meant relying off of them actually having the files available to download, in case they decide to pull a bedrock and make older versions of the game unavailable by removing their libraries and assets. They've done this in the past with older Office versions by pulling their installation media and updates as well as older Windows versions by pulling the installation media.

Additionally downloading the Minecraft library and asset files does not require authentication currently so it isn't piracy to download all the libraries and assets. It's only piracy if you use them in combination with a crack to play without an account.

Also PolyMC requires you to have a valid account to even download Minecraft game files, so doing this through PolyMC would not be piracy.

**Comment by DioEgizio** [comment]
> > > So that players can play any version of Minecraft in the future without having to rely on Microsoft.
> > 
> > 
> > wouldn't that be piracy?
> 
> By relying off of Microsoft I meant relying off of them actually having the files available to download, in case they decide to pull a bedrock and make older versions of the game unavailable by removing their libraries and assets. They've done this in the past with older Office versions by pulling their installation media and updates as well as older Windows versions by pulling the installation media.
> 
> Additionally downloading the Minecraft library and asset files does not require authentication currently so it isn't piracy to download all the libraries and assets. It's only piracy if you use them in combination with a crack to play without an account.
> 
> Also PolyMC requires you to have a valid account to even download Minecraft game files, so doing this through PolyMC would not be piracy.

oh then ok

**Comment by cozyGalvinism** [comment]
This should be a job for polymorphosis/metamorphosis instead of the launcher imo, as it's used for metadata anyways

**Comment by txtsd** [comment]
> This should be a job for polymorphosis/metamorphosis instead of the launcher imo, as it's used for metadata anyways

It would need to be stored in the instance folder in a structure usable by the game, which PolyMC already handles.

**Comment by cozyGalvinism** [comment]
PolyMC could dynamically link metamorphosis, though I'd have to extern the functions... dunno how well that'll work since it's async

**Comment by Idumein** [comment]
I made a program that can do this, you can find it [here](https://github.com/Idumein/mcdl).

However, my program downloads the game files from Mojang's server as-is, in the same directory structure, meaning it still needs to be sorted and placed in the correct folder in the launcher.

It should serve as no replacement for what a built-in option in PolyMC would be like, and I would like to encourage the implementation of such feature, however it is there in the meanwhile if anyone needs it asap.
